Add initial ppc64le support

madscientist159 opened the pull request:
Tested on Debian Bookworm

`resources/ubuntu-build-all.sh ~/depot_tools ~/WebRTC`

madscientist159 committed:
Add initial ppc64le support

JonathanLennox commented:
Is the goal here to be able to build natively on ppc64le?  As far as I know the cross-build from x86_64 to ppc64le is working -- is this also needed?

madscientist159 commented:
@JonathanLennox Yes, that is the goal.  amd64 hosts aren't permitted for critical services (either build or runtime) in our internal security model due to known security defects.

JonathanLennox commented:
Does the `USE_MAKEFILE` setting not work for you?

madscientist159 commented:
> Does the `USE_MAKEFILE` setting not work for you?

It does, this PR was prompted by a request in #6  I think the idea is that we don't want to be maintaining the `USE_MAKEFILE` path just for `ppc64le`.

fitzsim commented:
> > Does the `USE_MAKEFILE` setting not work for you?
> 
> It does, this PR was prompted by a request in #6 I think the idea is that we don't want to be maintaining the `USE_MAKEFILE` path just for `ppc64le`.

Right, that was my intention with the request; to eliminate maintenance of the separate `Makefile` and use upstream build steps on `ppc64le`.

fitzsim commented:
@madscientist159 thanks for posting this.

I fixed some README steps in #14.  Do they match how you're testing?

The first run, after having deleted `~/depot_tools` and `~/WebRTC`, fails for me:

```
$ resources/ubuntu-build-all.sh ~/depot_tools ~/WebRTC
[...]
[0:06:48] Still working on:
[0:06:48]   src/third_party
Syncing projects: 100% (63/63), done.             

________ running 'cipd ensure -log-level error -root /build/WebRTC -ensure-file /tmp/tmpqnr7uwa6.ensure' in '.'
Errors:
  failed to resolve gn/gn/linux-ppc64le@git_revision:635a71e20e99db0b08520ca0746b501167a5c7e5 (line 5): no such package: gn/gn/linux-ppc64le
  failed to resolve infra/rbe/client/linux-ppc64le@re_client_version:0.179.0.28341fc7-gomaip (line 8): no such package: infra/rbe/client/linux-ppc64le
  failed to resolve infra/3pp/tools/ninja/linux-ppc64le@version:3@1.12.1.chromium.4 (line 14): no such package: infra/3pp/tools/ninja/linux-ppc64le
  failed to resolve infra/build/siso/linux-ppc64le@git_revision:9d34ddf5122059c051fd4f52bf0150e93144e122 (line 21): no such package: infra/build/siso/linux-ppc64le
Error: Command 'cipd ensure -log-level error -root /build/WebRTC -ensure-file /tmp/tmpqnr7uwa6.ensure' returned non-zero exit status 1
Errors:
  failed to resolve gn/gn/linux-ppc64le@git_revision:635a71e20e99db0b08520ca0746b501167a5c7e5 (line 5): no such package: gn/gn/linux-ppc64le
  failed to resolve infra/rbe/client/linux-ppc64le@re_client_version:0.179.0.28341fc7-gomaip (line 8): no such package: infra/rbe/client/linux-ppc64le
  failed to resolve infra/3pp/tools/ninja/linux-ppc64le@version:3@1.12.1.chromium.4 (line 14): no such package: infra/3pp/tools/ninja/linux-ppc64le
  failed to resolve infra/build/siso/linux-ppc64le@git_revision:9d34ddf5122059c051fd4f52bf0150e93144e122 (line 21): no such package: infra/build/siso/linux-ppc64le

Subprocess failed with return code 1.
```
but immediately re-running the same command succeeds:
```
$ resources/ubuntu-build-all.sh ~/depot_tools ~/WebRTC
[...]
[100%] Linking CXX shared library libdcsctp4j.so
[100%] Built target dcsctp4j
Install the project...
-- Install configuration: "RelWithDebInfo"
-- Installing: /build/dcsctp4j/src/main/resources/linux-ppc64le/./libdcsctp4j.so
```
and the resulting snapshot jar file tests successfully on my server.

I tried both the current `main` branch and rebased onto the `7204` update pull request, and got the same results.

Are you able to test this approach when cross-compiling `ppc64le` binaries from an `x86-64` host?  @JonathanLennox are you maybe able to run this through an `x86-64` CI run to check?

That was another important motivation for writing the `Makefile`, since the `dcsctp4j` maintainers intend to cross-compile the `ppc64le` binaries for e.g., continuous integration.

madscientist159 commented:
> @madscientist159 thanks for posting this.
> 
> I fixed some README steps in #14. Do they match how you're testing?

Indeed, however I made a mistake and didn't include `VPYTHON_BYPASS` in the script.  We shouldn't be requiring the user to set that outside of the build scripts, I've fixed that in the current PR.

> The first run, after having deleted `~/depot_tools` and `~/WebRTC`, fails for me:
>
> but immediately re-running the same command succeeds:

Fixed.

> and the resulting snapshot jar file tests successfully on my server.
> 
> I tried both the current `main` branch and rebased onto the `7204` update pull request, and got the same results.
> 
> Are you able to test this approach when cross-compiling `ppc64le` binaries from an `x86-64` host? @JonathanLennox are you maybe able to run this through an `x86-64` CI run to check?

I don't have access to an `amd64` host of sufficient size / performance to handle anything Google -- the only hosts we have are left over open firmware systems from 2013 (AMD Family 15h) for a few legacy applications. 

> That was another important motivation for writing the `Makefile`, since the `dcsctp4j` maintainers intend to cross-compile the `ppc64le` binaries for e.g., continuous integration.

That makes sense.  I suspect the current scripting will fail, if so, what I'd do is update the scripts to check for both target and host, and if either one is `ppc64le` use the bypass method (system binaries vs. Google binaries).  Sound reasonable?

fitzsim commented:
> > @madscientist159 thanks for posting this.
> > I fixed some README steps in #14. Do they match how you're testing?
> 
> Indeed, however I made a mistake and didn't include `VPYTHON_BYPASS` in the script. We shouldn't be requiring the user to set that outside of the build scripts, I've fixed that in the current PR.

Makes sense.  I removed that step from #14.  I also kept track of all the packages I had to install on `x86-64`, and added those.

> > The first run, after having deleted `~/depot_tools` and `~/WebRTC`, fails for me:
> > but immediately re-running the same command succeeds:
> 
> Fixed.

Confirmed fixed (native ppc64le build).  Now I get another error, but I think this is because I rebased to #12 at the same time:

```
Skipping build for arch x86-64 on OpenPOWER host
Skipping build for arch arm64 on OpenPOWER host
ERROR at //.gn:38:25: Assignment had no effect.
exec_script_allowlist = build_dotfile_settings.exec_script_allowlist +
                        ^---------------------------------------------
You set the variable "exec_script_allowlist" here and it was unused before it went
out of scope.
```

I don't think this is worth debugging yet though (see below).

> I don't have access to an `amd64` host of sufficient size / performance to handle anything Google -- the only hosts we have are left over open firmware systems from 2013 (AMD Family 15h) for a few legacy applications.

I have a sufficient (`Libreboot`'ed, ME "disabled") `Debian 12` `x86-64` machine.  I was able to cross-compile a working `libdcsctp4j.so` on it, albeit using `powerpc64le-linux-gnu-gcc`, not `clang`, by adapting your extra `gn` arguments.  This approach can replace all the functionality provided by the `Makefile` with less maintenance.

> > That was another important motivation for writing the `Makefile`, since the `dcsctp4j` maintainers intend to cross-compile the `ppc64le` binaries for e.g., continuous integration.
> 
> That makes sense. I suspect the current scripting will fail, if so, what I'd do is update the scripts to check for both target and host, and if either one is `ppc64le` use the bypass method (system binaries vs. Google binaries). Sound reasonable?

I would like to phase in all these changes as follows:

1. Don't do anything until #12 is merged.
2. After that I will send in a pull request that replaces the `Makefile` with `gn` and `ninja` for `x86-64`-hosted `pp64le` builds, still using `GCC`.
3. I should probably do another pull request to change (2) to build with `clang` instead (like other architectures), if I can figure out how to set up `clang`-`x86-64`-host-to-`ppc64le`-cross compilation.
4. You can rebase the native `ppc64le` build support on (1,2, and 3).

(1) is because the `libwebrtc` repo update brings in some extra `ppc64le` support.

madscientist159 commented:
@fitzsim That seems reasonable to me.  I'll wait for your two PRs and rebase / test on top.

Thanks again for your work on this!  It'll be interesting to compare the two files produced from the native build and from the CI/CD cross build; ideally they should be binary identical but I doubt they will be.  The differences should be enlightening...
